ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,456 @@
+                panic!(
+                    "Unexpected query result: Expected 'Mapped', found '{:?}'",
+                    query_result
+                );
+            }
+        }
+    }
+
+    /// Get the mapped page.
+    /// This function will fail in the following cases:
+    /// * if the provided page type doesn't match the actual mapped one.
+    /// * If the address is not mapped (`NotMapped`), the function will fail.
+    /// * If the address is mapped to a `MappedUntracked` page, the function will fail.
+    pub fn get_page<T: PageMeta>(&self, addr: Vaddr) -> Result<Page<T>> {
+        assert!(self.start() <= addr && self.end() >= addr);
+        let start = addr.align_down(PAGE_SIZE);
+        let vaddr = start..start + PAGE_SIZE;
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let mut cursor = page_table.cursor(&vaddr).unwrap();
+        let query_result = cursor.query().unwrap();
+        match query_result {
+            PageTableItem::Mapped {
+                va: _,
+                page,
+                prop: _,
+            } => {
+                let result = Page::<T>::try_from(page);
+                if let Ok(page) = result {
+                    Ok(page)
+                } else {
+                    panic!("the provided page type doesn't match the actual mapped one");
+                }
+            }
+            _ => {
+                panic!(
+                    "Unexpected query result: Expected 'Mapped', found '{:?}'",
+                    query_result
+                );
+            }
+        }
+    }
+}
+
+/// The threshold used to determine whether we need to flush all TLB entries
+/// when handling a bunch of TLB flush requests. If the number of requests
+/// exceeds this threshold, the overhead incurred by flushing pages
+/// individually would surpass the overhead of flushing all entries at once.
+const TLB_FLUSH_ALL_THRESHOLD: usize = 32;
+
+cpu_local! {
+    /// The queue of pending requests.
+    static TLB_FLUSH_REQUESTS: SpinLock<VecDeque<TlbFlushRequest>> = SpinLock::new(VecDeque::new());
+}
+
+#[derive(Debug, Clone)]
+struct TlbFlushRequest {
+    op: TlbFlushOp,
+    // If we need to remove a mapped page from the page table, we can only
+    // recycle the page after all the relevant TLB entries in all CPUs are
+    // flushed. Otherwise if the page is recycled for other purposes, the user
+    // space program can still access the page through the TLB entries.
+    #[allow(dead_code)]
+    drop_after_flush: Option<DynPage>,
+}
+
+#[derive(Debug, Clone)]
+enum TlbFlushOp {
+    All,
+    Address(Vaddr),
+}
+
+impl TlbFlushRequest {
+    /// Perform the TLB flush operation on the current CPU.
+    fn do_flush(&self) {
+        use crate::arch::mm::{tlb_flush_addr, tlb_flush_all_excluding_global};
+        match &self.op {
+            TlbFlushOp::All => tlb_flush_all_excluding_global(),
+            TlbFlushOp::Address(addr) => tlb_flush_addr(*addr),
+        }
+    }
+}
+
```

Comment:
They are make public into `ostd::mm::tlb` in #1369 . After that is merged you can use these APIs so there's no need for wheels here.